proxy/coco/src/request/waiting_room.rs

```diff
@@ -119,16 +119,42 @@ impl<T> WaitingRoom<T> {
         )
     }
 
-    pub fn queried(&mut self, urn: &RadUrn, timestamp: T) -> Result<Request<IsRequested, T>, Error>
+    pub fn queried_requested(
+        &mut self,
+        urn: &RadUrn,
+        timestamp: T,
+    ) -> Result<Either<Request<TimedOut, T>, Request<IsRequested, T>>, Error>
     where
         T: Clone,
     {
+        let max_queries = self.max_queries;
+        let max_clones = self.max_clones;
         self.transition(
             |request| match request {
                 SomeRequest::Requested(request) => Some(request),
                 _ => None,
             },
-            |previous| previous.queried(timestamp),
+            |previous| previous.queried(max_queries, max_clones, timestamp),
+            urn,
+        )
+    }
+
+    pub fn queried_found(
+        &mut self,
```

Comment:
This one and the method above seem to assume knowledge on the caller side that implies state being kept per request. Rather whenever a `Have` is observed the caller should call `found`, if there is no request for it -- it's a noop otherwise the waiting room advances the request.